RATIONALE.md

```diff
@@ -1,5 +1,67 @@
 # Notable rationale of wazero
 
+## Project structure
+wazero uses internal packages extensively to balance API compatability desires for end users with the need to safely
+share internals between compilers.
+
+The end-user API includes the packages `wazero` `wasm` `wasi` with `Config` structs. Everything else is internal.
+
+### Internal packages
+Most code in wazero is internal, and it is acknowledged that this prevents external implementation of facets such as
+compilers or decoding. It also prevents splitting this code into separate repositories, resulting in a larger monorepo.
+This also adds work as more code needs to be centrally reviewed.
+
+However, the alternative is neither secure nor viable. To allow external implementation would require exporting symbols
+public, such as the `CodeSection`, which can easily create bugs. Moreover there's a high drift risk for any attempt at
+external implementations, compounded not just by wazero's code organization, but also the fast moving Wasm and WASI
+specifications.
+
+For example, implementing a compiler correctly requires expertise in Wasm, Golang and assembly. This requires deep
+insight into how internals are meant to be structured and the various tiers of testing required for `wazero` to result
+in a high quality experience. Even if someone had these skills, supporting external code would introduce variables which
+are constants in the central one. Supporting an external codebase is harder on the project team, and could starve time
+from the already large burden on the central codebase.
+
+The tradeoffs of internal packages are a larger codebase and responsibility to implement all standard features. It also
+implies thinking about extension more as forking is not viable for reasons above also. The primary mitigation of these
+realities are friendly OSS licensing, high rigor and a collaborative spirit which aim to make contribution in the shared
+codebase productive.
+
+### Avoiding cyclic dependencies
```

Comment:
maybe it's time to use the graph? https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/ :D